### PR TITLE
Merge master into nullable feature

### PR DESCRIPTION
**Pull request by safern**
This fixes merge conflicts (very simple ones) and pulls in the newest compiler.

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs [hunk 1]
```diff
@@ -915,16 +915,16 @@ internal static Task<TResult> GetTaskForResult(TResult result)
...
+                    (typeof(TResult) == typeof(uint) && default == (uint)(object?)result!) ||
+                    (typeof(TResult) == typeof(byte) && default(byte) == (byte)(object?)result!) ||
+                    (typeof(TResult) == typeof(sbyte) && default(sbyte) == (sbyte)(object?)result!) ||
+                    (typeof(TResult) == typeof(char) && default(char) == (char)(object?)result!) ||
+                    (typeof(TResult) == typeof(long) && default == (long)(object?)result!) ||
+                    (typeof(TResult) == typeof(ulong) && default == (ulong)(object?)result!) ||
+                    (typeof(TResult) == typeof(short) && default(short) == (short)(object?)result!) ||
+                    (typeof(TResult) == typeof(ushort) && default(ushort) == (ushort)(object?)result!) ||
+                    (typeof(TResult) == typeof(IntPtr) && default == (IntPtr)(object?)result!) ||
+                    (typeof(TResult) == typeof(UIntPtr) && default == (UIntPtr)(object?)result!)) // TODO-NULLABLE: https://github.com/dotnet/roslyn/issues/34976
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: Can we remove the question marks from object? now that we're adding the !s?

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs [hunk 2]
```diff
@@ -541,103 +541,103 @@ public bool GetBoolean(string name)
         {
             Type foundType;
             object? value = GetElement(name, out foundType);
-            return ReferenceEquals(foundType, typeof(bool)) ? (bool)value : _converter.ToBoolean(value!); // if value is null To* method will either deal with it or throw
+            return ReferenceEquals(foundType, typeof(bool)) ? (bool)value! : _converter.ToBoolean(value!); // if value is null To* method will either deal with it or throw
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: Could we add the `!` to `GetElement(…)` and make `value` be `object`?

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs [hunk 3]
```diff
@@ -541,103 +541,103 @@ public bool GetBoolean(string name)
...
         }
 
         public char GetChar(string name)
         {
             Type foundType;
             object? value = GetElement(name, out foundType);
-            return ReferenceEquals(foundType, typeof(char)) ? (char)value : _converter.ToChar(value!);
+            return ReferenceEquals(foundType, typeof(char)) ? (char)value! : _converter.ToChar(value!); // TODO-NULLABLE: https://github.com/dotnet/roslyn/issues/34976
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: Same question for all of these.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs (hunk 1 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: Actually yes... would it makes sense to change the typeof == typeof to use `is` instead? Or that would change IL?

**Comment by stephentoub** [comment]
I reviewed the last commit, which other than a couple comments LGTM.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: > Actually yes... would it makes sense to change the typeof == typeof to use is instead? Or that would change IL?

That would definitely change the IL (and resulting asm) in a bad way.  Please don't do that :)

Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: 😄 🤦‍♂️ ok.

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs (hunk 2 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: Good idea.

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs (hunk 3 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: Fixed.

File: src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs (hunk 2 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs]: Fixed.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs (hunk 1 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: Fixed.

**Pull request merged**